File: code/apps/web/src/lib/billing/account.ts
/**
 * The patient account as the cashier's desk reads it
 * (SRS-BIL-006, SRS-BIL-008, SRS-BIL-010, SRS-BIL-012, SRS-BIL-013, UX-W1-06).
 *
 * Two rules carry this screen.
 *
 * The balance derives from the ledger. SRS-BIL-012 says so, and the reason is
 * that a balance column kept in step by every writer drifts the first time one
 * of them fails between its two writes — with nothing to reconcile against,
 * because the ledger was never the truth. So the screen recomputes the balance
 * from the entries it was given and says so if that disagrees with the server's
 * figure, rather than displaying whichever arrived last.
 *
 * A finalised invoice is never edited. There is no edit control anywhere near
 * one (SRS-BIL-010). The patient is holding a copy, and a system that can
 * quietly change what a document said is a system whose documents prove
 * nothing. A correction is a credit or debit note that references the original,
 * which is what every finance department already does and every auditor expects
 * to find.
 */
import { addMoney, formatMoney, isZero, sumMoney, type Money } from './money.js';

/** Mirrors billing.v1.EntryKind. */
export type EntryKind =
	| 'invoice'
	| 'payment'
	| 'refund'
	| 'deposit'
	| 'deposit_applied'
	| 'write_off'
	| 'adjustment'
	| 'unspecified';

/** Mirrors billing.v1.InvoiceStatus. */
export type InvoiceStatus = 'draft' | 'issued' | 'superseded' | 'cancelled' | 'unspecified';

/** Mirrors billing.v1.ChargeStatus. */
export type ChargeStatus = 'posted' | 'invoiced' | 'voided' | 'held' | 'unspecified';

/** Mirrors billing.v1.PaymentMethod. */
export type PaymentMethod =
	| 'cash'
	| 'card'
	| 'upi'
	| 'bank_transfer'
	| 'cheque'
	| 'payer_settlement'
	| 'unspecified';

/** One ledger entry as the statement shows it. */
export interface PresentedEntry {
	readonly entryId: string;
	readonly kind: EntryKind;
	readonly kindLabel: string;
	readonly amount: Money;
	/** True when the entry increases what the patient owes. */
	readonly increasesDebt: boolean;
	readonly receiptNumber: string;
	readonly method: PaymentMethod;
	readonly methodLabel: string;
	readonly reason: string;
	readonly occurredAt: Date;
	readonly recordedBy: string;
}

const entryLabels: Record<EntryKind, string> = {
	invoice: 'Invoice',
	payment: 'Payment',
	refund: 'Refund',
	deposit: 'Deposit',
	// Distinct from a payment: the money was already held, and applying it is a
	// movement within the account rather than a new receipt.
	deposit_applied: 'Deposit applied',
	write_off: 'Write-off',
	adjustment: 'Adjustment',
	unspecified: 'Entry'
};

const methodLabels: Record<PaymentMethod, string> = {
	cash: 'Cash',
	card: 'Card',
	upi: 'UPI',
	bank_transfer: 'Bank transfer',
	cheque: 'Cheque',
	payer_settlement: 'Payer settlement',
	unspecified: ''
};

/** Human label for a ledger entry kind. */
export function describeEntryKind(kind: EntryKind): string {
	return entryLabels[kind];
}

/** Human label for a payment method. */
export function describeMethod(method: PaymentMethod): string {
	return methodLabels[method];
}

/** Presents one ledger entry. */
export function presentEntry(entry: {
	readonly entryId: string;
	readonly kind: EntryKind;
	readonly amount: Money;
	readonly receiptNumber: string;
	readonly method: PaymentMethod;
	readonly reason: string;
	readonly occurredAt: Date;
	readonly recordedBy: string;
}): PresentedEntry {
	return {
		entryId: entry.entryId,
		kind: entry.kind,
		kindLabel: entryLabels[entry.kind],
		amount: entry.amount,
		// Taken from the sign, not from the kind. The ledger is signed —
		// positive increases what the patient owes — and a screen that decided
		// direction from the kind would disagree with the ledger the first time
		// an adjustment went the other way.
		increasesDebt: entry.amount.minor > 0n,
		receiptNumber: entry.receiptNumber,
		method: entry.method,
		methodLabel: methodLabels[entry.method],
		reason: entry.reason,
		occurredAt: entry.occurredAt,
		recordedBy: entry.recordedBy
	};
}

/** What the statement header shows. */
export interface Statement {
	readonly entries: readonly PresentedEntry[];
	/** The balance this screen computed from the entries above. */
	readonly derivedBalance: Money;
	/** The balance the server reported. */
	readonly reportedBalance: Money;
	/**
	 * True when the two disagree.
	 *
	 * Surfaced rather than resolved. It means either the entry list is
	 * incomplete — a page boundary, most likely — or something is genuinely
	 * wrong; both are things a cashier should see before taking money against
	 * the figure.
	 */
	readonly reconciles: boolean;
	readonly deposits: Money;
}

/**
 * Builds the statement, deriving the balance rather than trusting one column.
 *
 * SRS-BIL-012's acceptance is "balance derives from ledger and reconciles", and
 * the second half is the interesting one: the derivation is only worth anything
 * if a disagreement is visible.
 */
export function buildStatement(params: {
	readonly entries: readonly PresentedEntry[];
	readonly reportedBalance: Money;
	readonly deposits: Money;
	readonly currency: string;
}): Statement {
	const derived = sumMoney(
		params.entries.map((entry) => entry.amount),
		params.currency
	);
	return {
		entries: [...params.entries].sort(
			(a, b) => b.occurredAt.getTime() - a.occurredAt.getTime()
		),
		derivedBalance: derived,
		reportedBalance: params.reportedBalance,
		reconciles: derived.minor === params.reportedBalance.minor,
		deposits: params.deposits
	};
}

/** An invoice as the list shows it. */
export interface PresentedInvoice {
	readonly invoiceId: string;
	readonly number: string;
	readonly status: InvoiceStatus;
	readonly statusLabel: string;
	readonly total: Money;
	readonly issuedAt: Date | null;
	/** Non-empty when a later document replaced this one. */
	readonly supersededBy: string;
	/** Non-empty when this document corrects an earlier one. */
	readonly correctsInvoiceId: string;
	/**
	 * False for anything finalised. There is no edit path on an issued invoice
	 * — a correction is a credit or debit note (SRS-BIL-010).
	 */
	readonly editable: boolean;
	/** True when a correction may be raised against it. */
	readonly correctable: boolean;
	readonly lineCount: number;
}

const invoiceLabels: Record<InvoiceStatus, string> = {
	draft: 'Draft',
	issued: 'Issued',
	// Retained and readable: the patient may be holding the superseded copy.
	superseded: 'Superseded',
	cancelled: 'Cancelled',
	unspecified: 'Unknown'
};

/** Human label for an invoice status. */
export function describeInvoiceStatus(status: InvoiceStatus): string {
	return invoiceLabels[status];
}

/** Presents one invoice. */
export function presentInvoice(invoice: {
	readonly invoiceId: string;
	readonly number: string;
	readonly status: InvoiceStatus;
	readonly total: Money;
	readonly issuedAt: Date | null;
	readonly supersededBy: string;
	readonly correctsInvoiceId: string;
	readonly lineCount: number;
}): PresentedInvoice {
	return {
		invoiceId: invoice.invoiceId,
		number: invoice.number,
		status: invoice.status,
		statusLabel: invoiceLabels[invoice.status],
		total: invoice.total,
		issuedAt: invoice.issuedAt,
		supersededBy: invoice.supersededBy,
		correctsInvoiceId: invoice.correctsInvoiceId,
		editable: invoice.status === 'draft',
		// A superseded or cancelled document is not corrected again; the
		// correction goes against whatever replaced it.
		correctable: invoice.status === 'issued',
		lineCount: invoice.lineCount
	};
}

/** A charge as the worklist shows it. */
export interface PresentedCharge {
	readonly chargeId: string;
	readonly description: string;
	readonly department: string;
	readonly quantity: number;
	readonly total: Money;
	readonly status: ChargeStatus;
	readonly statusLabel: string;
	/** True when the charge is still waiting to be invoiced. */
	readonly unbilled: boolean;
	/** True when a package covered it, so it appears at zero deliberately. */
	readonly covered: boolean;
	readonly coverageNote: string;
	readonly occurredAt: Date;
}

const chargeLabels: Record<ChargeStatus, string> = {
	posted: 'Unbilled',
	invoiced: 'Invoiced',
	// Retained rather than deleted: a voided charge is a service somebody
	// decided not to bill for, and that decision is part of the record.
	voided: 'Voided',
	held: 'On hold',
	unspecified: 'Unknown'
};

/** Human label for a charge status. */
export function describeChargeStatus(status: ChargeStatus): string {
	return chargeLabels[status];
}

/** Presents one charge. */
export function presentCharge(charge: {
	readonly chargeId: string;
	readonly description: string;
	readonly department: string;
	readonly quantity: number;
	readonly total: Money;
	readonly status: ChargeStatus;
	readonly covered: boolean;
	readonly coverageNote: string;
	readonly occurredAt: Date;
}): PresentedCharge {
	return {
		chargeId: charge.chargeId,
		description: charge.description,
		department: charge.department,
		quantity: charge.quantity,
		total: charge.total,
		status: charge.status,
		statusLabel: chargeLabels[charge.status],
		unbilled: charge.status === 'posted',
		covered: charge.covered,
		coverageNote: charge.coverageNote,
		occurredAt: charge.occurredAt
	};
}

/** Sums the charges that will appear on the next invoice. */
export function unbilledTotal(charges: readonly PresentedCharge[], currency: string): Money {
	return sumMoney(
		charges.filter((charge) => charge.unbilled).map((charge) => charge.total),
		currency
	);
}

/** One reason an account cannot be closed. */
export interface CloseException {
	readonly check: string;
	readonly detail: string;
	readonly amount: Money | null;
	readonly references: readonly string[];
}

/** Whether the account may be closed, and why not. */
export interface CloseReadiness {
	readonly ready: boolean;
	readonly exceptions: readonly CloseException[];
	readonly message: string;
}

/**
 * Reports whether the account may be closed (SRS-BIL-013).
 *
 * The exceptions come from the server and are listed, not summarised. "Cannot
 * close" with no list is a dead end for whoever is at the desk; the list is
 * what turns it into a task.
 */
export function closeReadiness(exceptions: readonly CloseException[]): CloseReadiness {
	if (exceptions.length === 0) {
		return { ready: true, exceptions: [], message: 'Everything is settled. The account can be closed.' };
	}
	return {
		ready: false,
		exceptions,
		message:
			exceptions.length === 1
				? 'One thing is outstanding before this account can be closed.'
				: `${exceptions.length} things are outstanding before this account can be closed.`
	};
}

/** Why a payment cannot be taken. */
export interface PaymentValidity {
	readonly ready: boolean;
	readonly problems: readonly string[];
}

/**
 * Checks a payment before taking it.
 *
 * The idempotency key is required rather than generated at submit time. A key
 * minted on the click changes on every retry, so a double submission produces
 * two receipts — the whole point of the key is that it is stable across the
 * retry (SRS-BIL-008).
 */
export function validatePayment(payment: {
	readonly amount: Money | null;
	readonly method: PaymentMethod;
	readonly idempotencyKey: string;
	readonly accountId: string;
}): PaymentValidity {
	const problems: string[] = [];

	if (payment.accountId.trim() === '') {
		problems.push('No account is open.');
	}
	if (!payment.amount) {
		// parseMoney returned null: what was typed could not be read as an
		// amount, and refusing is the only safe answer.
		problems.push('Enter an amount.');
	} else if (payment.amount.minor <= 0n) {
		// A refund is its own operation against the original payment
		// (SRS-BIL-009), not a negative payment.
		problems.push('A payment must be more than zero. Use a refund to return money.');
	}
	if (payment.method === 'unspecified') {
		problems.push('Choose how the money was taken.');
	}
	if (payment.idempotencyKey.trim() === '') {
		problems.push('This payment has no idempotency key.');
	}

	return { ready: problems.length === 0, problems };
}

/** A one-line summary of what the desk owes or holds. */
export function describeAccountHeader(statement: Statement, locale?: string): string {
	const parts: string[] = [];
	if (!isZero(statement.deposits)) {
		parts.push(`${formatMoney(statement.deposits, { locale })} on deposit`);
	}
	const balance = statement.reconciles
		? statement.reportedBalance
		: // When the two disagree, the derived figure is shown, because it is the
			// one this screen can account for line by line.
			statement.derivedBalance;
	parts.push(
		balance.minor > 0n
			? `${formatMoney(balance, { locale })} outstanding`
			: balance.minor < 0n
				? `${formatMoney({ minor: -balance.minor, currency: balance.currency }, { locale })} in credit`
				: 'settled'
	);
	return parts.join(', ');
}

/** Adds a single entry to a running balance, for an optimistic update. */
export function applyEntry(balance: Money, entry: PresentedEntry): Money {
	return addMoney(balance, entry.amount);
}

File: code/apps/mobile/lib/src/billing/account.dart
/// The patient account as the cashier's desk reads it
/// (UX-W1-06, SRS-BIL-006, SRS-BIL-008, SRS-BIL-010, SRS-BIL-012, SRS-BIL-013).
///
/// Two rules carry this screen.
///
/// The balance derives from the ledger. SRS-BIL-012 says so, and the reason is
/// that a balance column kept in step by every writer drifts the first time one
/// of them fails between its two writes — with nothing to reconcile against,
/// because the ledger was never the truth. So the screen recomputes the balance
/// from the entries it was given and says so when that disagrees with the
/// server's figure, rather than displaying whichever arrived last.
///
/// A finalised invoice is never edited. There is no edit control anywhere near
/// one (SRS-BIL-010). The patient is holding a copy, and a system that can
/// quietly change what a document said is a system whose documents prove
/// nothing. A correction is a credit or debit note referencing the original,
/// which is what every finance department already does and every auditor
/// expects to find.
library;

import 'package:meta/meta.dart';

import 'money.dart';

/// Mirrors billing.v1.EntryKind.
enum EntryKind {
  invoice,
  payment,
  refund,
  deposit,
  depositApplied,
  writeOff,
  adjustment,
  unspecified,
  unrecognised,
}

/// Mirrors billing.v1.InvoiceStatus.
enum InvoiceStatus {
  draft,
  issued,
  superseded,
  cancelled,
  unspecified,
  unrecognised,
}

/// Mirrors billing.v1.ChargeStatus.
enum ChargeStatus { posted, invoiced, voided, held, unspecified, unrecognised }

/// Mirrors billing.v1.PaymentMethod.
enum PaymentMethod {
  cash,
  card,
  upi,
  bankTransfer,
  cheque,
  payerSettlement,
  unspecified,
  unrecognised,
}

String describeEntryKind(EntryKind kind) => switch (kind) {
      EntryKind.invoice => 'Invoice',
      EntryKind.payment => 'Payment',
      EntryKind.refund => 'Refund',
      EntryKind.deposit => 'Deposit',
      EntryKind.depositApplied => 'Deposit applied',
      EntryKind.writeOff => 'Written off',
      EntryKind.adjustment => 'Adjustment',
      EntryKind.unspecified => 'Entry',
      EntryKind.unrecognised => 'Entry not recognised by this app',
    };

String describeMethod(PaymentMethod method) => switch (method) {
      PaymentMethod.cash => 'Cash',
      PaymentMethod.card => 'Card',
      PaymentMethod.bankTransfer => 'Bank transfer',
      PaymentMethod.upi => 'UPI',
      PaymentMethod.cheque => 'Cheque',
      // Not "insurance": the money came from a payer settling a claim, which
      // is a different event from a patient paying with an insurance card.
      PaymentMethod.payerSettlement => 'Payer settlement',
      PaymentMethod.unspecified => 'Not recorded',
      PaymentMethod.unrecognised => 'Method not recognised by this app',
    };

String describeInvoiceStatus(InvoiceStatus status) => switch (status) {
      InvoiceStatus.draft => 'Draft',
      InvoiceStatus.issued => 'Issued',
      InvoiceStatus.superseded => 'Superseded',
      InvoiceStatus.cancelled => 'Cancelled',
      InvoiceStatus.unspecified => 'Unknown',
      InvoiceStatus.unrecognised => 'Status not recognised by this app',
    };

String describeChargeStatus(ChargeStatus status) => switch (status) {
      ChargeStatus.posted => 'Posted',
      ChargeStatus.invoiced => 'Invoiced',
      ChargeStatus.voided => 'Voided',
      ChargeStatus.held => 'Held',
      ChargeStatus.unspecified => 'Unknown',
      ChargeStatus.unrecognised => 'Status not recognised by this app',
    };

/// One ledger entry as the statement shows it.
@immutable
class PresentedEntry {
  const PresentedEntry({
    required this.entryId,
    required this.kind,
    required this.amount,
    required this.occurredAt,
    required this.increasesDebt,
    this.receiptNumber = '',
    this.method = PaymentMethod.unspecified,
    this.reason = '',
    this.recordedBy = '',
  });

  final String entryId;
  final EntryKind kind;
  final Money amount;
  final DateTime occurredAt;

  /// Taken from the sign, not from the kind. The ledger is signed — positive
  /// increases what the patient owes — and a screen that decided direction
  /// from the kind would disagree with the ledger the first time an adjustment
  /// went the other way.
  final bool increasesDebt;

  final String receiptNumber;
  final PaymentMethod method;
  final String reason;
  final String recordedBy;

  String get kindLabel => describeEntryKind(kind);
  String get methodLabel => describeMethod(method);
}

/// Presents one ledger entry.
PresentedEntry presentEntry({
  required String entryId,
  required EntryKind kind,
  required Money amount,
  required DateTime occurredAt,
  String receiptNumber = '',
  PaymentMethod method = PaymentMethod.unspecified,
  String reason = '',
  String recordedBy = '',
}) =>
    PresentedEntry(
      entryId: entryId,
      kind: kind,
      amount: amount,
      occurredAt: occurredAt,
      increasesDebt: amount.minor > 0,
      receiptNumber: receiptNumber,
      method: method,
      reason: reason,
      recordedBy: recordedBy,
    );

/// What the statement header shows.
@immutable
class Statement {
  const Statement({
    required this.entries,
    required this.derivedBalance,
    required this.reportedBalance,
    required this.reconciles,
    required this.deposits,
  });

  final List<PresentedEntry> entries;

  /// The balance this screen computed from the entries above.
  final Money derivedBalance;

  /// The balance the server reported.
  final Money reportedBalance;

  /// True when the two agree.
  ///
  /// A disagreement is surfaced rather than resolved. It means either the entry
  /// list is incomplete — a page boundary, most likely — or something is
  /// genuinely wrong; both are things a cashier should see before taking money
  /// against the figure.
  final bool reconciles;

  final Money deposits;
}

/// Builds the statement, deriving the balance rather than trusting one column.
///
/// SRS-BIL-012's acceptance is "balance derives from ledger and reconciles",
/// and the second half is the interesting one: the derivation is only worth
/// anything if a disagreement is visible.
Statement buildStatement({
  required List<PresentedEntry> entries,
  required Money reportedBalance,
  required Money deposits,
  required String currency,
}) {
  final derived = sumMoney(
    [for (final entry in entries) entry.amount],
    currency,
  );
  return Statement(
    entries: [...entries]..sort((a, b) => b.occurredAt.compareTo(a.occurredAt)),
    derivedBalance: derived,
    reportedBalance: reportedBalance,
    reconciles: derived.minor == reportedBalance.minor,
    deposits: deposits,
  );
}

/// An invoice as the list shows it.
@immutable
class PresentedInvoice {
  const PresentedInvoice({
    required this.invoiceId,
    required this.number,
    required this.status,
    required this.total,
    required this.editable,
    required this.correctable,
    this.issuedAt,
    this.supersededBy = '',
    this.correctsInvoiceId = '',
    this.lineCount = 0,
  });

  final String invoiceId;
  final String number;
  final InvoiceStatus status;
  final Money total;
  final DateTime? issuedAt;
  final String supersededBy;
  final String correctsInvoiceId;

  /// Only ever true for a draft (SRS-BIL-010).
  final bool editable;

  /// True when a credit or debit note may reference this one.
  final bool correctable;

  final int lineCount;

  String get statusLabel => describeInvoiceStatus(status);
}

/// Presents one invoice.
PresentedInvoice presentInvoice({
  required String invoiceId,
  required String number,
  required InvoiceStatus status,
  required Money total,
  DateTime? issuedAt,
  String supersededBy = '',
  String correctsInvoiceId = '',
  int lineCount = 0,
}) =>
    PresentedInvoice(
      invoiceId: invoiceId,
      number: number,
      status: status,
      total: total,
      issuedAt: issuedAt,
      supersededBy: supersededBy,
      correctsInvoiceId: correctsInvoiceId,
      // Draft and only draft. An issued invoice is in the patient's hand.
      editable: status == InvoiceStatus.draft,
      // A superseded or cancelled document is not corrected again; the
      // correction goes against whatever replaced it.
      correctable: status == InvoiceStatus.issued,
      lineCount: lineCount,
    );

/// A charge as the worklist shows it.
@immutable
class PresentedCharge {
  const PresentedCharge({
    required this.chargeId,
    required this.display,
    required this.status,
    required this.total,
    required this.unbilled,
    this.occurredAt,
  });

  final String chargeId;
  final String display;
  final ChargeStatus status;
  final Money total;

  /// True when this charge will appear on the next invoice.
  ///
  /// Not "is not yet on an invoice", which is a wider set and the mistake this
  /// field invites: a held charge is also not on an invoice, and it is held
  /// precisely so that it does not reach one until a coding query or an
  /// authorisation is resolved. Counting it inflates a figure somebody bills
  /// from and quotes to a patient.
  final bool unbilled;

  /// True when the charge is waiting on a coding query or an authorisation.
  ///
  /// Its own flag so the desk can show it — a held charge is actionable, and
  /// dropping it off the screen entirely hides work somebody has to chase.
  bool get held => status == ChargeStatus.held;

  final DateTime? occurredAt;

  String get statusLabel => describeChargeStatus(status);
}

/// Presents one charge.
PresentedCharge presentCharge({
  required String chargeId,
  required String display,
  required ChargeStatus status,
  required Money total,
  DateTime? occurredAt,
}) =>
    PresentedCharge(
      chargeId: chargeId,
      display: display,
      status: status,
      total: total,
      // Posted only. Invoiced is already on a document, voided is not a charge
      // any more, and held is waiting on a query — see the field's own comment
      // for why that last one is the trap.
      unbilled: status == ChargeStatus.posted,
      occurredAt: occurredAt,
    );

/// Sums the charges not yet on an invoice.
Money unbilledTotal(List<PresentedCharge> charges, String currency) => sumMoney(
      [for (final c in charges) if (c.unbilled) c.total],
      currency,
    );

/// One reason an account cannot be closed.
@immutable
class CloseException {
  const CloseException({
    required this.check,
    required this.detail,
    this.amount,
    this.references = const [],
  });

  final String check;
  final String detail;
  final Money? amount;
  final List<String> references;
}

/// Whether the account may be closed, and why not.
@immutable
class CloseReadiness {
  const CloseReadiness({
    required this.ready,
    required this.exceptions,
    required this.message,
  });

  final bool ready;
  final List<CloseException> exceptions;
  final String message;
}

/// Reports whether the account may be closed (SRS-BIL-013).
///
/// The exceptions come from the server and are listed, not summarised. "Cannot
/// close" with no list is a dead end for whoever is at the desk; the list is
/// what turns it into a task.
CloseReadiness closeReadiness(List<CloseException> exceptions) {
  if (exceptions.isEmpty) {
    return const CloseReadiness(
      ready: true,
      exceptions: [],
      message: 'Everything is settled. The account can be closed.',
    );
  }
  return CloseReadiness(
    ready: false,
    exceptions: exceptions,
    message: exceptions.length == 1
        ? 'One thing is outstanding before this account can be closed.'
        : '${exceptions.length} things are outstanding before this account '
            'can be closed.',
  );
}

/// Why a payment cannot be taken.
@immutable
class PaymentValidity {
  const PaymentValidity({required this.ready, required this.problems});
  final bool ready;
  final List<String> problems;
}

/// Checks a payment before taking it.
///
/// The idempotency key is required rather than generated at submit time. A key
/// minted on the click changes on every retry, so a double submission produces
/// two receipts — the whole point of the key is that it is stable across the
/// retry (SRS-BIL-008).
PaymentValidity validatePayment({
  required Money? amount,
  required PaymentMethod method,
  required String idempotencyKey,
  required String accountId,
}) {
  final problems = <String>[
    if (accountId.trim().isEmpty) 'No account is open.',
    // A null amount means parseMoney refused what was typed. Refusing is the
    // only safe answer: a lenient parser turns a typing slip into zero.
    if (amount == null)
      'Enter an amount.'
    else if (amount.minor <= 0)
      // A refund is its own operation against the original payment
      // (SRS-BIL-009), not a negative payment.
      'A payment must be more than zero. Use a refund to return money.',
    if (method == PaymentMethod.unspecified ||
        method == PaymentMethod.unrecognised)
      'Choose how the money was taken.',
    if (idempotencyKey.trim().isEmpty)
      'This payment has no idempotency key.',
  ];

  return PaymentValidity(ready: problems.isEmpty, problems: problems);
}

/// A one-line summary of what the desk owes or holds.
String describeAccountHeader(
  Statement statement, {
  MoneyGrouping grouping = MoneyGrouping.thousands,
}) {
  final parts = <String>[
    if (!isZero(statement.deposits))
      '${formatMoney(statement.deposits, grouping: grouping)} on deposit',
  ];

  // When the two disagree, the derived figure is shown, because it is the one
  // this screen can account for line by line.
  final balance =
      statement.reconciles ? statement.reportedBalance : statement.derivedBalance;

  parts.add(
    balance.minor > 0
        ? '${formatMoney(balance, grouping: grouping)} outstanding'
        : balance.minor < 0
            ? '${formatMoney(Money(minor: -balance.minor, currency: balance.currency), grouping: grouping)} in credit'
            : 'settled',
  );
  return parts.join(', ');
}

/// Adds a single entry to a running balance, for an optimistic update.
Money applyEntry(Money balance, PresentedEntry entry) =>
    addMoney(balance, entry.amount);
